File: pipeline/updateRegistry.js
#!/usr/bin/env node
/**
 * pipeline/updateRegistry.js
 *
 * Rebuilds registry/exercises.json from data/pipeline-status.json.
 * Run after uploading new videos to a CDN (R2, Cloudinary, etc.)
 * or after fetching new Pexels CDN URLs.
 *
 * Usage:
 *   node pipeline/updateRegistry.js
 *
 * Also accepts R2 inventory via environment:
 *   R2_PUBLIC_URL=https://pub-xxx.r2.dev node pipeline/updateRegistry.js
 */

import { readFileSync, writeFileSync, mkdirSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, "..");

const STATUS_FILE = join(ROOT, "data", "pipeline-status.json");
const REGISTRY_FILE = join(ROOT, "registry", "exercises.json");
const TERMS_FILE = join(ROOT, "data", "video_search_terms.json");

// Determine category from the slug → known exercises vs osteopathy
function getCategory(slug, terms) {
  const exerciseSlugs = new Set(terms.exercises.map((e) => e.id));
  return exerciseSlugs.has(slug) ? "exercises" : "osteopathy";
}

// Load existing registry (if any) as base
function loadExisting() {
  try {
    return JSON.parse(readFileSync(REGISTRY_FILE, "utf-8"));
  } catch {
    return {};
  }
}

function main() {
  // Load pipeline-status.json
  let statusData = {};
  try {
    statusData = JSON.parse(readFileSync(STATUS_FILE, "utf-8"));
  } catch {
    console.error("ERROR: data/pipeline-status.json not found. Run fetch_pexels_urls.py first.");
    process.exit(1);
  }

  // Load video_search_terms to determine categories
  let terms = { exercises: [], osteopathy: [] };
  try {
    terms = JSON.parse(readFileSync(TERMS_FILE, "utf-8"));
  } catch {
    console.warn("WARN: data/video_search_terms.json not found — categories will be 'unknown'");
  }

  const existing = loadExisting();
  const registry = { ...existing };

  let added = 0;
  let updated = 0;

  for (const [slug, entry] of Object.entries(statusData)) {
    const video = entry.video;
    if (!video) continue;

    const category = getCategory(slug, terms);
    const current = registry[slug];

    if (!current) {
      registry[slug] = { video, category };
      added++;
    } else if (current.video !== video) {
      registry[slug] = { ...current, video, category };
      updated++;
    }
  }

  mkdirSync(join(ROOT, "registry"), { recursive: true });
  writeFileSync(REGISTRY_FILE, JSON.stringify(registry, null, 2));

  const total = Object.keys(registry).length;
  console.log(`✓ registry/exercises.json updated`);
  console.log(`  Total entries: ${total}`);
  console.log(`  Added: ${added}, Updated: ${updated}, Unchanged: ${total - added - updated}`);
  console.log(`  CDN: ${Object.values(registry).filter((e) => e.video?.startsWith("https://")).length} entries with CDN URLs`);
  console.log(`  Local: ${Object.values(registry).filter((e) => e.video?.startsWith("/")).length} entries with local paths`);
}

main();

File: pipeline/update_registry.py
#!/usr/bin/env python3
"""
pipeline/update_registry.py
Patches data/master-registry.json with cloudflare_id and video_url
from pipeline/generation_status.json.

Run after upload_to_cloudflare.py, then git push to trigger Netlify rebuild.

Usage:
  python3 pipeline/update_registry.py
  git add data/master-registry.json
  git commit -m "feat: update registry with Cloudflare Stream video IDs"
  git push
"""

import json
from pathlib import Path

# ─── Paths ────────────────────────────────────────────────────────────────────
BASE_DIR        = Path(__file__).parent.parent
STATUS_PATH     = Path(__file__).parent / "generation_status.json"
REGISTRY_PATH   = BASE_DIR / "data" / "master-registry.json"

# ─── Helpers ──────────────────────────────────────────────────────────────────
def load_json(path: Path):
    with open(path, encoding="utf-8") as f:
        return json.load(f)

def save_json(path: Path, data):
    with open(path, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=2, ensure_ascii=False)

# ─── Main ─────────────────────────────────────────────────────────────────────
def main():
    if not STATUS_PATH.exists():
        print("ERROR: pipeline/generation_status.json not found.")
        return

    status   = load_json(STATUS_PATH)
    registry = load_json(REGISTRY_PATH)

    updated  = 0
    skipped  = 0
    missing  = 0

    for registry_id, entry in status.items():
        cf_id = entry.get("cloudflare_id", "")

        if not cf_id:
            skipped += 1
            continue

        if registry_id not in registry:
            missing += 1
            print(f"  Warning: {registry_id} in status but not in master-registry.json")
            continue

        registry[registry_id]["cloudflare_id"] = cf_id
        registry[registry_id]["video_url"] = f"https://watch.videodelivery.net/{cf_id}"
        updated += 1

    save_json(REGISTRY_PATH, registry)

    total_with_video = sum(1 for v in registry.values() if v.get("cloudflare_id"))

    print(f"Therapia Global — Registry Updater")
    print(f"Updated: {updated} entries")
    print(f"Skipped (no cloudflare_id yet): {skipped}")
    print(f"Missing in registry: {missing}")
    print(f"Total with video: {total_with_video} / {len(registry)}")
    print()
    print("data/master-registry.json saved.")
    print()
    print("Deploy:")
    print("  git add data/master-registry.json")
    print(f"  git commit -m 'feat: {updated} Cloudflare Stream video IDs added'")
    print("  git push")
    print()
    print("Netlify will rebuild automatically. Videos will be live in ~2 minutes.")

if __name__ == "__main__":
    main()
